### PR TITLE
DOC: Bayesian estimation of SARIMAX using PyMC3 NUTS

### PR DESCRIPTION
**Pull request by aeturrell**
<details>
Added example showing how to do Bayesian estimation of SARIMAX models and unobserved components models. Put together by Chad Fulton and myself.
</details>

### CODE REVIEW

**Comment by josef-pkt** [comment]
Just from the maintainer side:
We are not adding the content with the notebooks, i.e. the output should be cleared (not be in any commit of the PR)

The main reason is that we don't want to increase the memory size of the repo and the time for checkouts with things that can be autogenerated, and need to be regenerated if something changes.
(e.g. We don't want a git history that contains hundreds of mostly obsolete pictures.)

**Comment by josef-pkt** [comment]
for the workflow: 
I usually use a gist during development of a notebook.
Alternatively, it's possible to keep the notebook output in a branch during development, but then squash it before merging with an empty notebook in the commit.

**Comment by aeturrell** [comment]
Sounds completely sensible, thanks - so, to confirm, I just need to redo the commit with no output in the notebook?

**Comment by josef-pkt** [comment]
> so, to confirm, I just need to redo the commit with no output in the notebook?

Yes

However, you can also postpone that, so @ChadFulton can look at the complete notebook
https://github.com/statsmodels/statsmodels/blob/fb0f3c4d9c099bb4ce7dd1412c4f67e823c6a3b1/examples/notebooks/bayesian_est_SARIMAX.ipynb

You could also open a new PR with the empty notebook, when review here is finished.
(we don't get any interesting git history from notebooks, so squashing/merging individual commits doesn't destroy useful information)

**Comment by ChadFulton** [comment]
Great to see a PR with this notebook @aeturrell, thanks!

I think the content of the notebook is great (not surprising since you already showed it to me last year :) ).  For the record, it basically shows how to use the state space models for Bayesian estimation using the NUTS sampler from PyMC3, which is pretty neat.

There are just a couple of things that we need to investigate:

1. As @josef-pkt says, our notebooks are typically stored as unevaluated .ipynb files, which are then run by the Travis (I think) job when the documentation is generated. However:
    1. This notebook requires `theano` and `pymc3` packages to run. @bashtage do you happen to know if this is a problem with our documentation generation?
    2. This notebook may take a while to run, since it does the MCMC sampling.  Looking at the PyMC3 output, it looks like it's at least 3-4 minutes runtime. This seems like a problem for the usual setup?
    3. (I have another notebook in the CFA PR, #6501, that has similar issues to these)
2. (minor) I think we need to increase the indentation levels of all the headings by one, since I think the documentation requires that the notebook start one-level down from the top. So we should just add an extra `#` to each of the headings.

**Comment by aeturrell** [comment]
Thanks Chad. Obviously you massively helped with the notebook. While you and Josef deliberate on the dependency and auto-generation issue, what's most useful for me to do? Is it possible for me to modify the pull request so that I can make the two changes: a) remove output from notebook and b) fix indentation? Or will I need to create a new PR on a fresh branch? Apologies - I'm new to this!

**Comment by bashtage** [comment]
You need to add both theano and pymc3 to this file:

https://github.com/statsmodels/statsmodels/blob/a34eb5d5f77eb73679232476105f3876ca3fddee/tools/ci/docbuild_install.sh#L11

I think you will need to use conda since they do not appear to have binaries available on PyPI

```
# Add pymc3 and theano for bayesian_est_sarimax.
echo conda install theano pymc3 -y
conda install theano pymc3 -y
```

The current error is:

```
******************************************************************************
ERROR: Error occurred when running /home/travis/build/statsmodels/statsmodels/examples/notebooks/bayesian_est_SARIMAX.ipynb
An error occurred while executing the following cell:
------------------
import theano
import theano.tensor as tt
import pymc3 as pm
import numpy as np
import matplotlib.pyplot as plt
import statsmodels.api as sm
import pandas as pd
from pandas_datareader.data import DataReader
plt.style.use('seaborn')
from pandas.plotting import register_matplotlib_converters
register_matplotlib_converters()
------------------
---------------------------------------------------------------------------
ModuleNotFoundError                       Traceback (most recent call last)
<ipython-input-1-d67c0d6dca5d> in <module>
----> 1 import theano
      2 import theano.tensor as tt
      3 import pymc3 as pm
      4 import numpy as np
      5 import matplotlib.pyplot as plt
ModuleNotFoundError: No module named 'theano'
ModuleNotFoundError: No module named 'theano'
An error occurred while executing the following cell:
------------------
import theano
import theano.tensor as tt
import pymc3 as pm
import numpy as np
import matplotlib.pyplot as plt
import statsmodels.api as sm
import pandas as pd
from pandas_datareader.data import DataReader
plt.style.use('seaborn')
from pandas.plotting import register_matplotlib_converters
register_matplotlib_converters()
------------------
---------------------------------------------------------------------------
ModuleNotFoundError                       Traceback (most recent call last)
<ipython-input-1-d67c0d6dca5d> in <module>
----> 1 import theano
      2 import theano.tensor as tt
      3 import pymc3 as pm
      4 import numpy as np
      5 import matplotlib.pyplot as plt
ModuleNotFoundError: No module named 'theano'
ModuleNotFoundError: No module named 'theano'
******************************************************************************
```

**Comment by bashtage** [comment]
I would also vote for renaming the noteboook to look like the others, e.g, statespace_sarimax_bayesian_estimation or something along these lines.

**Comment by bashtage** [comment]
Need a `%matplotlib inline` to silence the warnings I think.

**Comment by ChadFulton** [comment]
Thanks very much @bashtage!

@aeturrell it looks like you won't need to create a new branch.  I think we need to do the following:

1. Add `%matplotlib inline` as the very first line in the first Python cell of the notebook (i.e. the line just before `import theano`)
1. Clear the output from the notebook (If you haven't cleared output before, select "Cell > All Output > Clear" in the notebook and then save it).
1. Rename the notebook. Personally, I would suggest using `statespace_sarimax_pymc3.ipynb`. For consistency, we may as well rename the notebook's image file as well (and then modify `landing.yml` accordingly).
1. Modify the file `statsmodels/tools/ci/docbuild_install.sh` to include the lines @bashtage suggested:
    ```bash
    # Add pymc3 and theano for bayesian_est_sarimax.
    echo conda install theano pymc3 -y
    conda install theano pymc3 -y
    ```
1. I notice that the notebook's image file is bigger than necessary - I usually use 720 x 450. So to save some space, we might want to replace it with a smaller file. In case it helps, I reduced the resolution to 720 x 450 and the resulting image is at the bottom of this comment.

Thanks again for doing the work to get this up!

statespace_sarimax_pymc3

**Comment by ChadFulton** [comment]
Sorry, and then once you've done those steps, just make a new commit locally and then `git push` the commit up to the same repository as before. This PR will automatically update.

**Comment by aeturrell** [comment]
Thanks all for the clear advice. Really helpful. Let me know if there are other changes needed.

**Comment by ChadFulton** [comment]
Thanks!

One more thing - since Statsmodels' master branch had additional commits made after you created your branch, you ended up merging those changes into your branch as a new commit. You can see the last commit in this PR is "Merge branch 'master' into bayesian_est_sarimax". This is problematic though because it makes the Git history hard to follow.  Instead, what we want to do is to move your commits to the end of the chain of commits currently in master, which you do with a rebase rather than a merge (for lots more details, see e.g. https://hackernoon.com/git-merge-vs-rebase-whats-the-diff-76413c117333).

I think you should be able to fix this by doing:

```bash
git rebase <remote_repo>/master
git push --force
```

where `<remote_repo>` is whatever you named the Statsmodels repository that you're pulling from, and you can probably get that from:

```bash
git remote -v
```

**Comment by aeturrell** [comment]
Thanks Chad. I'm not sure that worked. Here's what I did and got:
`git remote -v
origin	https://github.com/aeturrell/statsmodels.git (fetch)
origin	https://github.com/aeturrell/statsmodels.git (push)
upstream	https://github.com/statsmodels/statsmodels (fetch)
upstream	https://github.com/statsmodels/statsmodels (push)`

I couldn't find any combination of those that would work as `<remote_repo>' part. Looking at the documentation, eg [https://git-scm.com/book/en/v2/Git-Branching-Rebasing](here), it seemed to suggest I should be doing
`git rebase master'
so I checked out the master branch, git pulled to get the latest version, then did `git rebase master' before `git push --force'. But then there were conflicts (in `landing.yml') which does imply something wasn't right.

**Comment by bashtage** [comment]
```
git checkout <your-branch>
git remote add upstream https://github.com/statsmodels/statsmodels.git
git fetch --all
git rebase upstream/master
```

If everything goes well, you can finally `git push -f`

**Comment by aeturrell** [comment]
Thanks Kevin, that's super clear and helpful. Hopefully all resolved now.

**Comment by bashtage** [comment]
Thanks.  LGTM. We need to take it in so that we can see it on the site.  There is a small chance that a follow-up might be required.

**Comment by Sandy4321** [comment]
but what is about Bayesian classifier for categorical data , especially when we have observations in train data that not present in test data 
scikit implementation can not deal with this 
https://scikit-learn.org/stable/modules/generated/sklearn.naive_bayes.CategoricalNB.html#sklearn.naive_bayes.CategoricalNB

**Comment by bashtage** [comment]
@Sandy4321 That doesn't seem to have anything to do with Bayesian estimation of a SARIMAX model.

**Comment by Sandy4321** [comment]
but what is about Bayesian classifier for categorical data , especially when we have observations in train data that not present in test data
scikit implementation can not deal with this
https://scikit-learn.org/stable/modules/generated/sklearn.naive_bayes.CategoricalNB.html#sklearn.naive_bayes.CategoricalNB
